File: frontend/lib/data/models/product/product_model.dart
class ProductModel {
  String? category;
  String? title;
  String? description;
  int? price;
  List<String>? images;
  String? sId;
  String? updatedOn;
  String? createdOn;

  ProductModel(
      {this.category,
      this.title,
      this.description,
      this.price,
      this.images,
      this.sId,
      this.updatedOn,
      this.createdOn});

  ProductModel.fromJson(Map<String, dynamic> json) {
    category = json['category'];
    title = json['title'];
    description = json['description'];
    price = json['price'];
    images = json['images'].cast<String>();
    sId = json['_id'];
    updatedOn = json['updatedOn'];
    createdOn = json['createdOn'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    data['category'] = this.category;
    data['title'] = this.title;
    data['description'] = this.description;
    data['price'] = this.price;
    data['images'] = this.images;
    data['_id'] = this.sId;
    data['updatedOn'] = this.updatedOn;
    data['createdOn'] = this.createdOn;
    return data;
  }
}

File: backend/src/models/product_model.js
const { Schema, model } = require('mongoose');
//create category model
const productSchema = new Schema({
    category: { type: Schema.Types.ObjectId, ref: 'Category', required: [true, 'category is required'] },
    title: { type: String, required: [true, 'title is required'] },
    description: { type: String, default: '' },
    price: { type: Number, required: true, },
    images: { type: Array, default: [] },
    updatedOn: { type: Date },
    createdOn: { type: Date },
});

// pre and update product model
productSchema.pre('save', function (next) {
    this.updatedOn = new Date();
    this.createdOn = new Date();
    next();
})
productSchema.pre(['update', 'updateOne', 'findOneAndUpdate'], function (next) {
    const update = this.getUpdate();
    delete update._id;
    this.updatedOn = new Date;
    next();
})



//export category model
const productModel = model('Product', productSchema);
module.exports = productModel;
